### DISCUSSION TITLE
Notifications vs. Communication

### DISCUSSION THREAD

**Discussion by theyosh** [theyosh/TerrariumPI#597] [Ideas]
Hi all,

As TP4 is basically ready, the only remaining thing is the Notifications part. And it turns out a bit of more work and I am not able to test all the social media connections. I stopped using Telegram, Pusher, Twitter etc. So I am unable to test those notifications in TP4.

Also, there are always requests regarding the Notifications to add certain/more triggers and logic. And I spent more time on the Notifications part then the actual goal: more logic for the terrariums and other enclosures.

And after a few years, new API and clouds are being used. So that means even more code for the Notifications part. This is getting to much.

Therefore, I am thinking of changing the Notifications part to Communications and add less options. For now I am thinking of supporting:
- HTTP(S) POST REST API
- MQTT
- Email
- Local script
- Matrix (optional)
- Trafficlight (optional)
- Display (optional)

With the above options, you are able to make your own integration with other services. For example, with Matrix you could setup your own bridge to Telegram, Signal, Slack, etc..
And this way, I do not have to think of logic for which sensor you would receive a notification and which sensor not. With the communications you will get the callback for every sensor, and then using your own logic, you decide what to do with that information.

So in short:
- Notifications become Communication
- Simple logic in TP4 for just on/off certain communication
- As a owner you need to create some small pieces of software for further integration

**Comment by ntfd0** [comment]
Nice! Sounds great I think, solves a lot of hassle down the road for sure

**Comment by theyosh** [comment]
So,

How to handle MQTT? You can create topic for each sensor, relay, webcam etc. But is that the right way.
Now, I have topics in the format: `terrariumpi/sensor/` and `terrariumpi/relay/` etc.
Where there is a sensor ID and a unique ID so that you can pick your messages you want. I think this is easier then a topic per sensor like `terrariumpi/sensor/[SENSOR_ID]/`
And the data is then all the data that is available as a JSON message. So the message is plaintext and you can parse it to a JSON object.

Also I will not install a MQTT server. For testing I am using HiveMQ.com which is free for low amount of data. But I guess any MQTT server will do. So I leave that for the person that want to use MQTT. I can support SSL connection.

And the last thing is how to describe the topics and data that is available. Any suggestions anybody?

**Comment by ntfd0** [comment]
@theyosh Sorry for a late reply!
So the idea is that all data from ALL sensors will be sent out as a single JSON message, and then we will need to parse that on our own...?
How would that work with different enclosures? (sorry if I misunderstood anything)
Using a third-party MQTT server should be okay, theoretically, but it'd be nice to have an out-of-the-box solution. Priority-wise, having our own MQTT server is definitely not very urgent and/or important.

As for the topics, I've been thinking about the following:
- Temp sensors (separate by enclosure):
   -  Over threshold, below threshhold (for each sensor separately)
   - Avg temperature (as a periodic message perhaps?)
   - Whenever a temperature reading triggers another event (e.g. temp too high, turning off the relay controlling a heater)
   - Incorrect reading (to indicate that a sensor is broken or disconnected)
- Humidity sensors (also separate by enclosure)
   - Over threshold, below threshhold (for each enclosure separately should be enough? by sensor is fine too)
   - Avg humidity (as a periodic message perhaps?)
   - Whenever a humidity reading triggers another event (e.g. humidity too low, turning on the relay controlling a mister)
   - Incorrect reading (to indicate that a sensor is broken or disconnected)
- Door sensors (separate by enclosure)
   - Door open for more than X seconds (make it customisable perhaps? I have garter and corn snakes who are excellent escape artists, so for me this would be like 30 seconds max, I think)
- Relays (separate by enclosure)
   - Relay triggered
That's about the extent of the functionality I'm currently using. Others to include might be the water level.

**Reply by theyosh** [reply]
Hmm, the idea with mqtt is that I would have less logic on my side. But I am not sure if that works with the mqtt design patterns: https://hackernoon.com/a-detailed-guide-to-the-world-of-mqtt-bo1d63cay

So I was thinking that sensors just send out the total object as data. So something like a topic: `terrariumpi/sensors/[id/name]/` or humidity, etc, depending on the sensor type. For alarm I could do: `terrariumpi/sensors/[id/name]/alarm/` with again all info.
So I was thinking of sending all data that is needed to make your own conclusion. But that is again the correct mqtt patten. Also you cannot use a simple client to show the data.

So, I thought that it was quiet easy to implement mqtt, and technical I have it running already, but I/we need to think a bit more about the topics and which information to send. My way of doing is a bit abusing the mqtt service... :(

You example of door and alarm after 30 seconds is a nice example. But that needs a setting for 'how long' is to long. Where to enter that? It is possible at the creating of the enclosure below the door selection. But then again, this is more a feature to have a alarm when the doors are open to long. And that has nothing to do with mqtt itself. So here we have some mixing with logic and a way of sending out the alarm.

Need some more discussion and thoughts

**Comment by kebs19** [comment]
I find it essential to have a lcd display to show me that the system is working, the temps are ok, etc. 
I use it to check the hours for more than once, I admit 😄

**Reply by kebs19** [reply]
Pushover is a must-have for me, I run the system on a Pi Zero and from time to time it stops working and I have to reboot. 
I use Pushover to receive the error messages from TerrariumPI to know when the system is facing issues.

**Reply by theyosh** [reply]
Hmm, Pushover was on the drop list, but the integration is that simple, that I have added it to TP4. If you are able to test, that would be nice.

Make sure you have the latest code `git pull` and rerun the installer. As new libraries needs to be installed.

**Reply by kebs19** [reply]
hey there! 
glad to test it :D 

I did the git pull, reinstall it and set up a pushover notification with api and id keys, it worked until I set a new notification to message me when a relay changed state (so I can test if the system is sending me the notifications) and then the whole system stoped.
It doesn't load at all. 
When I debug it, I have the following msg:

Traceback (most recent call last):
  File "terrariumPI.py", line 16, in <module>
    terrariumEngine = terrariumEngine(__version__)
  File "/home/pi/TerrariumPI/terrariumEngine.py", line 133, in __init__
    self.__load_existing_relays()
  File "/home/pi/TerrariumPI/terrariumEngine.py", line 647, in __load_existing_relays
    self.relays[relay.id].set_state(last_value, True)
  File "/home/pi/TerrariumPI/hardware/relay/__init__.py", line 231, in set_state
    self.callback(self.id, self.state)
  File "/home/pi/TerrariumPI/terrariumEngine.py", line 748, in callback_relay
    self.notification.message(f'relay_change' , relay_data)
  File "/home/pi/TerrariumPI/terrariumNotification.py", line 154, in message
    self.services[service.id].send_message(message_id, title, text, data)
  File "/home/pi/TerrariumPI/terrariumNotification.py", line 845, in send_message
    r = requests.post(self.setup['url'],
KeyError: 'url'

**Reply by kebs19** [reply]
I opened the DB file with DB Browser and deleted all the entries in "NotificationMessage" table. upload it back to the PI and it started to work again (without the notifications). I'll leave it like that and wait for new instructions

**Reply by theyosh** [reply]
Ok, I just made a stupid typo. So do a new `git pull` and restart. I think it should work then. And thanks for testing!

**Reply by kebs19** [reply]
I did the git pull and now the pages are loading correctly, but when a notification event happens it gives the following error:

2021-10-06 18:49:09,442 - ERROR   - terrariumNotification - Error sending Pusover message 'teste' with status code: 400

**Reply by theyosh** [reply]
Made an update. So do a `git pull` and restart. I hope it should work now.

I am also thinking of making a some kind of a test. So you do not need to wait for an event to test.

**Reply by kebs19** [reply]
I did the git pull, but I'm still receiving the same error:
2021-10-09 19:57:55,669 - ERROR   - terrariumNotification - Error sending Pusover message 'teste' with status code: 400

Another issue I found is when I try to delete a notification message (in this case a relay-change notification) I get the following error:

Error:
Area with id relay_change does not exists.

(Interesting that I can update the notification message, enable/disable it and get an OK answer, but I can't delete it)

**Reply by kebs19** [reply]
I tried to create a new notif msg, but it doesn't alert me (and give no error message also). 
I made a 'login error' event, and the system can see the event:
2021-10-09 20:18:52,071 - WARNING - terrariumWebserver    - Incorrect login detected using username 'admin222222' and password 'password' from ip 192.168.1.11

But can't alert me when it happened.

**Reply by theyosh** [reply]
Found the issue. Apparently I did have a paid Pushover account :P So I could test it myself... and it is working now....

Do a `git pull` and it should start working for you.
